sharding/simulator/service.go

```diff
@@ -0,0 +1,118 @@
+}
+
+// NewSimulator creates a struct instance of a simulator service.
+// It will have access to config, a mainchain client, a p2p server,
+// and a shardID.
+func NewSimulator(config *params.Config, client *mainchain.SMCClient, p2p *p2p.Server, shardID int, delay time.Duration) (*Simulator, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	errChan := make(chan error)
+	requestSent := make(chan interface{})
+	return &Simulator{config, client, p2p, shardID, ctx, cancel, errChan, requestSent, delay}, nil
+}
+
+// Start the main loop for simulating p2p requests.
+func (s *Simulator) Start() {
+	log.Info("Starting simulator service")
+	feed := s.p2p.Feed(messages.CollationBodyRequest{})
+	go s.simulateNotaryRequests(s.client.SMCCaller(), s.client.ChainReader(), feed)
+	go s.handleServiceErrors()
+}
+
+// Stop the main loop for simulator requests.
+func (s *Simulator) Stop() error {
+	// Triggers a cancel call in the service's context which shuts down every goroutine
+	// in this service.
+	defer s.cancel()
+	log.Warn("Stopping simulator service")
+	return nil
+}
+
+// handleServiceErrors manages a goroutine that listens for errors broadcast to
+// this service's error channel. This serves as a final step for error logging
+// and is stopped upon the service shutting down.
+func (s *Simulator) handleServiceErrors() {
+	for {
+		select {
+		case <-s.ctx.Done():
+			return
+		case err := <-s.errChan:
+			log.Error(fmt.Sprintf("Simulator service error: %v", err))
+		}
+	}
+}
+
+// simulateNotaryRequests simulates p2p message sent out by notaries
+// once the system is in production. Notaries will be performing
+// this action within their own service when they are selected on a shard, period
+// pair to perform their responsibilities. This function in particular simulates
+// requests for collation bodies that will be relayed to the appropriate proposer
+// by the p2p feed layer.
+func (s *Simulator) simulateNotaryRequests(fetcher mainchain.RecordFetcher, reader mainchain.Reader, feed *event.Feed) {
+	for {
+		select {
+		// Makes sure to close this goroutine when the service stops.
+		case <-s.ctx.Done():
+			return
+		case <-time.After(time.Second * s.delay):
+			blockNumber, err := reader.BlockByNumber(s.ctx, nil)
+			if err != nil {
+				s.errChan <- fmt.Errorf("could not fetch current block number: %v", err)
+				continue
+			}
+
+			period := new(big.Int).Div(blockNumber.Number(), big.NewInt(s.config.PeriodLength))
+			req, err := syncer.RequestCollationBody(fetcher, big.NewInt(int64(s.shardID)), period)
+			if err != nil {
+				s.errChan <- fmt.Errorf("error constructing collation body request: %v", err)
+				continue
+			}
+			if req != nil {
+				msg := p2p.Message{
+					Peer: p2p.Peer{},
+					Data: *req,
+				}
+				feed.Send(msg)
+
+				// Notifies the requestSent channel for any other handlers that could run upon
+				// this event occurring (also useful for tests.)
+				s.requestSent <- msg
```

Comment:
Were you able to verify that this does not block the log message below if no one is listening to `s.requestSent`?
I was having this issue yesterday.